### PR TITLE
Partially replace cleanup with finalizers

### PR DESCRIPTION
**Pull request by weyfonk**
This implements [finalizers](https://book.kubebuilder.io/reference/using-finalizers) on GitRepos, bundles and bundle deployments, removing the corresponding cleanup controller logic for those resources. That cleanup is no longer necessary, as those resources can no longer be orphaned, ie deleted when no Fleet controller is running.

This is tested by end-to-end tests, by temporarily scaling down the (unsharded) `fleet-controller` deployment and checking that deleting one of those resources has no effect other than setting their deletion timestamps.

In general, adding and removing finalizers on a resource is handled by that resource's reconciler, with the following exception: the `GitRepo` reconciler takes care of removing bundle deployment finalizers, when purging bundle deployments. This prevents the agent from dealing with finalizers, which would require a change in permissions and thereby a possible security risk.

### Out of scope

This does not add finalizers to RBAC resources created for GitRepos (eg. service accounts, roles, role bindings, config maps), as those are native k8s resources which can be created in more use cases than GitRepo lifecycle management (eg. agent management), therefore  cleanup logic is still needed there.

Content resources do not bear finalizers either, because cleaning them up responds to a different process according to reference counts, ie how many bundles refer to a given Content resource.

Refers to #1980

### CODE REVIEW

**Review by 0xavi0 — COMMENTED** [review]
Just a couple of questions. And, aside from preventing the user to delete `bundledeployments` I don't see why we need to set a finalizer in those.

In the version without finalisers the user is able to delete a `bundledeployment` which is is almos instantly recreated by Fleet. 

With this the user can't delete `bundledeployments` as the DeleteTimestamp is set but the finaliser is never removed. 
I don't know if keeping the `bundledeployment` with the delete timestamp set and alive is the right thing to do. 

There is no problem deleting `bundles` or `gitrepos` (I've tested also those cases with this branch)

File: e2e/single-cluster/finalizers_test.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+		_, _ = k.Delete("bundle", fmt.Sprintf("%s-%s", gitrepoName, path))
+		bdNamespace, err := k.Get(
+			"ns",
+			"-o=jsonpath='{.items[?(@."+
+				`metadata.annotations.fleet\.cattle\.io/cluster-namespace=="fleet-local"`+
+				")].metadata.name}'",
+		)
+		Expect(err).ToNot(HaveOccurred())
+
+		_, _ = env.Kubectl.Namespace(bdNamespace).Delete(
```
Comment (0xavi0) [inline e2e/single-cluster/finalizers_test.go:56]: Why do we need to "manually" delete the `bundledeployment`?
It should be gone after deleting the `bundle`, right?

In fact this should be part of the test... checking that when deleting a `bundle` or `gitrepo` the `bundledeployments` related are also gone

File: e2e/single-cluster/finalizers_test.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+				"--replicas=1",
+				"--timeout=5s",
+			)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = k.Delete("gitrepo", gitrepoName)
+			Expect(err).NotTo(HaveOccurred())
+
+			// These resources should be deleted when the GitRepo is deleted.
+			By("checking that the auxiliary resources don't exist anymore")
```
Comment (0xavi0) [inline e2e/single-cluster/finalizers_test.go:154]: We could also check that the `bundle`s and `bundledeployment`s are deleted

File: internal/cmd/controller/reconciler/bundle_controller.go [hunk 3]
```diff
@@ -150,6 +189,11 @@ func (r *BundleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		// NOTE we don't use the existing BundleDeployment, we discard annotations, status, etc
 		// copy labels from Bundle as they might have changed
 		bd := target.BundleDeployment()
+
+		if bd.DeletionTimestamp.IsZero() {
+			controllerutil.AddFinalizer(bd, bundleDeploymentFinalizer)
+		}
```
Comment (0xavi0) [inline internal/cmd/controller/reconciler/bundle_controller.go:195]: In what case is this going to be different to zero?
In my understanding the `BundleDeployment` struct was just instantiated and it's still not added to the cluster.

I guess that we could omit the `if` statement if that's correct

**Review by 0xavi0 — COMMENTED** [review]
Just a couple of questions. And, aside from preventing the user to delete `bundledeployments` I don't see why we need to set a finalizer in those.

In the version without finalisers the user is able to delete a `bundledeployment` which is is almos instantly recreated by Fleet. 

With this the user can't delete `bundledeployments` as the DeleteTimestamp is set but the finaliser is never removed. 
I don't know if keeping the `bundledeployment` with the delete timestamp set and alive is the right thing to do. 

There is no problem deleting `bundles` or `gitrepos` (I've tested also those cases with this branch)

File: e2e/single-cluster/finalizers_test.go (hunk 2 above)
Comment (weyfonk) [reply e2e/single-cluster/finalizers_test.go:154]: This does not depend on finalizers though, but rather on relationships between a `GitRepo`, bundles and bundle deployments. We cover some of it [here](https://github.com/rancher/fleet/blob/main/e2e/single-cluster/gitrepo_test.go#L78).

File: internal/cmd/controller/reconciler/bundle_controller.go (hunk 3 above)
Comment (weyfonk) [reply internal/cmd/controller/reconciler/bundle_controller.go]: You are right, that condition was here as a precaution; I have replaced it with a comment explaining precisely that.

File: e2e/single-cluster/finalizers_test.go (hunk 1 above)
Comment (weyfonk) [reply e2e/single-cluster/finalizers_test.go]: I have removed this, as the bundle deployment will indeed have been deleted by each test case, either manually or through previous cleanup steps, ie scaling the Fleet controller back up, which would trigger `GitRepo` and bundle deletion.

**Review by 0xavi0 — APPROVED** [review]

**Review by 0xavi0 — APPROVED** [review]

**Pull request merged**